File: scala/aoc/src/main/scala/adventofcode/Day20/Day20.scala
package adventofcode.Day20

import scala.io.Source
import scala.util.Using
import scala.annotation.tailrec

type Point = (Int, Int)
type Image = Vector[Vector[Char]]
type Algorithm = Vector[Char]

def parsePixel(c: Char): Char =
  c match
    case '.' => '0'
    case '#' => '1'

def parse(input: String): (Algorithm, Image) =
  input.split("""\n\n""") match
    case Array(alg, img) => (
      alg.toVector.map(parsePixel),
      img.split("""\n""").toVector.map(line => line.toVector.map(parsePixel)))

def getWindow(p: Point): List[Point] =
  val (r, c) = p
  List((r-1, c-1), (r-1,c), (r-1,c+1), (r, c-1), (r,c), (r, c+1), (r+1, c-1), (r+1, c), (r+1, c+1))

def getValue(image: Image, p: Point, default: Char): Char =
  val (r,c) = p
  image.lift(r).flatMap(_.lift(c)).getOrElse(default)

def computeNewPixel(p: Point, image: Image, alg: Algorithm, default: Char): Char =
  val windowValues = getWindow(p).map(x => getValue(image, x, default))
  val algorithmIndex = Integer.parseInt(windowValues.mkString, 2)
  alg(algorithmIndex)

def nextRound(image: Image, algorithm: Algorithm, default: Char): Image =
  val numRows = image.length
  val numCols = image(0).length
  Vector.range(-1, numRows + 1)
    .map(r => Vector.range(-1, numCols + 1)
      .map(c => computeNewPixel((r,c), image, algorithm, default)))

def getDefaultValue(iteration: Int): Char =
  if iteration % 2 == 0 then '0' else '1'

def simulate(image: Image, algorithm: Algorithm, steps: Int): Image =
  @tailrec
  def loop(im: Image, n: Int): Image =
    if n == steps then im
    else loop(nextRound(im, algorithm, getDefaultValue(n)), n+1)
  loop(image, 0)

def countPixels(image: Image): Int =
  image.flatten.count(_ == '1')

@main def day20(): Unit =
  val input = Using.resource(Source.fromFile("input/day20.txt"))(_.mkString)
  val (algorithm, image) = parse(input)

  val part1 = countPixels(simulate(image, algorithm,2))
  val part2 = countPixels(simulate(image, algorithm,50))
  println(part1)
  println(part2)

File: clojure/aoc/src/aoc/day20.clj
(ns aoc.day20
  (:require [clojure.string :as str]))

(defn parse-pixel [c]
  (if (= c \.) \0 \1))

(defn parse [input]
  (let [[algorithm _ & image-string] (str/split-lines input)
        image-vec (mapv (fn [s] (mapv (fn [c] (parse-pixel c)) s)) image-string)]
    [(mapv parse-pixel algorithm) image-vec]))

(defn get-default-value [iteration]
  (if (= 0 (mod iteration 2)) \0 \1))

(defn get-window [[r c]]
  [[(dec r) (dec c)] [(dec r) c] [(dec r) (inc c)]
   [r (dec c)] [r c] [r (inc c)]
   [(inc r) (dec c)] [(inc r) c] [(inc r) (inc c)]])

(defn get-value [image value [r c]]
  (get-in image [r c] value))

(defn bin->dec [binary-string]
  (Integer/parseInt binary-string 2))

(defn next-round [image alg value]
  (mapv (fn [r] (mapv (fn [c]
                        (->> [r c]
                             get-window
                             (map #(get-value image value %))
                             (apply str)
                             bin->dec
                             alg)) (range -1 (-> image first count inc)))) (range -1 (-> image count inc))))

(defn simulate [starting-image alg rounds]
  (reduce (fn [image iteration] (next-round image alg (get-default-value iteration))) starting-image (range 0 rounds)))

(defn count-pixels [image]
    (->> (flatten image)
         (filter #(= \1 %))
         count))

(defn -main []
  (let [[alg image] (parse (slurp "resources/day20.txt"))
        part1 (-> (simulate image alg 2) count-pixels)
        part2 (-> (simulate image alg 50) count-pixels)]
    (run! println [part1 part2])))